src/world/GameWorldView.cpp

```diff
@@ -66,10 +69,54 @@ const GameWorldBase& GameWorldView::GetWorld() const
     return gwv.GetWorld();
 }
 
-void GameWorldView::SetZoomFactor(float zoomFactor)
+void GameWorldView::SetNextZoomFactor()
 {
-    this->zoomFactor = zoomFactor;
+    if (zoomFactor == targetZoomFactor)
+        return;
+
+    float zoomDirection = targetZoomFactor - zoomFactor;
+
+    // last step
+    if (abs(zoomDirection) < abs(zoomSpeed))
+    {
+        zoomFactor = targetZoomFactor;
+        zoomSpeed = 0;
+    }
+    // deceleration towards zero zoom speed
+    else if (abs(zoomDirection) <= 0.5 * zoomSpeed * zoomSpeed / ZOOM_ACCELERATION)
+    {
+        if (zoomSpeed > 0)
+            zoomSpeed -= ZOOM_ACCELERATION;
+        else
+            zoomSpeed += ZOOM_ACCELERATION;
+    }
+    // acceleration to unlimited speed
+    else
+    {
+        if (zoomDirection > 0)
+            zoomSpeed += ZOOM_ACCELERATION;
+        else
+            zoomSpeed -= ZOOM_ACCELERATION;
+    }
+
+    zoomFactor = zoomFactor + zoomSpeed;
     CalcFxLx();
+
+    // allow overshoot, but swing back
```

Comment:
I'd disallow this and add this check in SetZoomFactor. Don't think this "swing back" looks nice and it might even annoy users ("doesn't work!!!")